nibabel/streamlines/trk.py

```diff
@@ -0,0 +1,737 @@
+
+        # Update the 'scalar_name' field using 'data_per_point' of the tractogram.
+        data_for_points = first_item.data_for_points
+        if len(data_for_points) > MAX_NB_NAMED_SCALARS_PER_POINT:
+            raise ValueError("Can only store {0} named data_per_point (scalars).".format(MAX_NB_NAMED_SCALARS_PER_POINT))
+
+        data_for_points_keys = sorted(data_for_points.keys())
+        self.header['scalar_name'] = np.zeros(MAX_NB_NAMED_SCALARS_PER_POINT, dtype='S20')
+        for i, k in enumerate(data_for_points_keys):
+            nb_values = data_for_points[k].shape[1]
+
+            if len(k) > 20:
+                raise ValueError("Scalar name '{0}' is too long (max 18 char.)".format(k))
+            elif len(k) > 18 and nb_values > 1:
+                raise ValueError("Scalar name '{0}' is too long (need to be less than 18 characters when storing more than one value".format(k))
+
+            scalar_name = k
+            if nb_values > 1:
+                # Use the last to bytes of the name to store the nb of values associated to this data_for_streamline.
+                scalar_name = asbytes(k[:18].ljust(18, '\x00')) + b'\x00' + np.array(nb_values, dtype=np.int8).tostring()
+
+            self.header['scalar_name'][i] = scalar_name
+
+        # `Tractogram` streamlines are in RAS+ and mm space, we will compute
+        # the affine matrix that will bring them back to 'voxelmm' as required
+        # by the TRK format.
+        affine = np.eye(4)
+
+        # Applied the inverse of the affine found in the TRK header.
+        # rasmm -> voxel
+        affine = np.dot(np.linalg.inv(self.header[Field.VOXEL_TO_RASMM]), affine)
+
+        # If the voxel order implied by the affine does not match the voxel
+        # order in the TRK header, change the orientation.
+        # voxel (affine) -> voxel (header)
+        header_ornt = asstr(self.header[Field.VOXEL_ORDER])
+        affine_ornt = "".join(nib.orientations.aff2axcodes(self.header[Field.VOXEL_TO_RASMM]))
+        header_ornt = nib.orientations.axcodes2ornt(header_ornt)
+        affine_ornt = nib.orientations.axcodes2ornt(affine_ornt)
+        ornt = nib.orientations.ornt_transform(affine_ornt, header_ornt)
+        M = nib.orientations.inv_ornt_aff(ornt, self.header[Field.DIMENSIONS])
+        affine = np.dot(M, affine)
+
+        # TrackVis considers coordinate (0,0,0) to be the corner of the
+        # voxel whereas `Tractogram` streamlines assume (0,0,0) is the
+        # center of the voxel. Thus, streamlines are shifted of half a voxel.
+        offset = np.eye(4)
+        offset[:-1, -1] += 0.5
+        affine = np.dot(offset, affine)
+
+        # Finally send the streamlines in mm space.
+        # voxel -> voxelmm
+        scale = np.eye(4)
+        scale[range(3), range(3)] *= self.header[Field.VOXEL_SIZES]
+        affine = np.dot(scale, affine)
+
+        # The TRK format uses float32 as the data type for points.
+        affine = affine.astype(np.float32)
```

Comment:
Are streamlines also saved as float32? See https://github.com/nipy/dipy/issues/837 if iut might help fix the debate.